### PR TITLE
Handle pre-shared invite keys

### PR DESCRIPTION
**Pull request by AndrewFerr**
Signed-off-by: Andrew Ferrazzutti <andrewf@element.io>

## Checklist

* [ ] Tests written for all new code
* [x] Linter has been satisfied
* [x] Sign-off given on the changes (see CONTRIBUTING.md)

### CODE REVIEW

File: src/e2ee/RustEngine.ts [hunk 1]
```diff
@@ -90,9 +89,11 @@ export class RustEngine {
                 break;
             case "joined":
             default:
-            // Default and other cases handled by assignment before switch
+                memberships.splice(-1);
         }
 
+        const members = (await this.client.getRoomMembers(roomId, null, memberships)).map(u => new UserId(u.membershipFor));
```
Comment (turt2live) [inline src/e2ee/RustEngine.ts:95]: we should probably be using `getRoomMembersByMembership` so the set of invited users can fail without failing encryption.

**Review by turt2live — CHANGES_REQUESTED** [review]
the other thing I'm not sure on is whether this is actually recommended by the spec/rust-sdk - it doesn't appear to come up in documentation, though I'm willing to admit I might be blind.

**Comment by AndrewFerr** [comment]
> the other thing I'm not sure on is whether this is actually recommended by the spec/rust-sdk - it doesn't appear to come up in documentation, though I'm willing to admit I might be blind.

FWIW this is the behaviour Element Web uses.  If you've invited a user with an encryption-enabled device to an encrypted room, EW will send the room key to that user (via a to-device message) even before they've joined the room.  In fact, it sends the key even before you post a message -- it sends it as soon as you starting typing a message!

File: src/e2ee/RustEngine.ts [hunk 2]
```diff
@@ -90,9 +89,11 @@ export class RustEngine {
                 break;
             case "joined":
             default:
-            // Default and other cases handled by assignment before switch
+                memberships.splice(-1);
```
[the remark this answers is not in the document]
Comment (AndrewFerr) [reply src/e2ee/RustEngine.ts]: 7a2a3ff

File: src/e2ee/RustEngine.ts (hunk 1 above)
Comment (AndrewFerr) [reply src/e2ee/RustEngine.ts]: ccde51b

File: src/e2ee/RustEngine.ts [hunk 3]
```diff
@@ -90,7 +90,21 @@ export class RustEngine {
...
+        }
+
+        const members = new Set<UserId>();
+        for (const membership of memberships) {
+            try {
+                (await this.client.getRoomMembersByMembership(roomId, membership))
+                    .map(u => new UserId(u.membershipFor))
+                    .forEach(u => void members.add(u));
+            } catch (err) {
+                LogService.warn("RustEngine", `Failed to get room members for membership type "${membership}" in ${roomId}`, extractRequestError(err));
```
Comment (turt2live) [inline src/e2ee/RustEngine.ts:103]: I'm not sure we should be consuming this error tbh

File: src/e2ee/RustEngine.ts [hunk 4]
```diff
@@ -90,7 +90,21 @@ export class RustEngine {
...
-            // Default and other cases handled by assignment before switch
+                memberships = ["join"];
+        }
+
+        const members = new Set<UserId>();
+        for (const membership of memberships) {
+            try {
+                (await this.client.getRoomMembersByMembership(roomId, membership))
+                    .map(u => new UserId(u.membershipFor))
+                    .forEach(u => void members.add(u));
```
Comment (turt2live) [inline src/e2ee/RustEngine.ts:101]: preferably we use a real `for` loop for this, as one-liners mixed in with try/catch is a bit confusing imo

Comment (turt2live) [reply src/e2ee/RustEngine.ts:101]: also, we shouldn't need a `Set` here as someone can't exist in two states. Appending to an array (using `push(...userIds)`) should be fine?

**Pull request closed without merging**